### PR TITLE
SI-2823 Implicit in for expression

### PR DESCRIPTION
**Pull request by som-snytt**
Support for `implicit` keyword in front of generators and midstream assignments.

The optional modifier is an attachment to the pattern tree for purposes of desugaring.

```
scala> object X { def unapply(s: String): Option[(String,Int)] = { val x = s.split("/").head ; Option(x, x.length) } }
defined object X

scala> for (implicit X(s, i) <- List("ab/bbc", "xxx/y", "ab/c")) yield implicitly[String] + implicitly[Int] // show
object $read extends scala.AnyRef {
  def <init>() = {
    super.<init>;
    ()
  };
  object $iw extends scala.AnyRef {
    def <init>() = {
      super.<init>;
      ()
    };
    import $line3.$read.$iw.$iw.X;
    object $iw extends scala.AnyRef {
      def <init>() = {
        super.<init>;
        ()
      };
      val res0 = List("ab/bbc", "xxx/y", "ab/c").withFilter(((check$ifrefutable$1) => check$ifrefutable$1: @scala.unchecked match {
        case X((s @ _), (i @ _)) => true
        case _ => false
      })).map(((x$1) => {
        <synthetic> <artifact> private[this] val x$2 = x$1: @scala.unchecked: @scala.unchecked match {
          case X((s @ _), (i @ _)) => scala.Tuple2(s, i)
        };
        implicit val s = x$2._1;
        implicit val i = x$2._2;
        implicitly[String].+(implicitly[Int])
      }))
    }
  }
}
res0: List[String] = List(ab2, xxx3, ab2)

scala> for (X(s, i) <- List("ab/bbc", "xxx/y", "ab/c") ; implicit t = s) yield implicitly[String] + i // show
object $read extends scala.AnyRef {
  def <init>() = {
    super.<init>;
    ()
  };
  object $iw extends scala.AnyRef {
    def <init>() = {
      super.<init>;
      ()
    };
    import $line3.$read.$iw.$iw.X;
    object $iw extends scala.AnyRef {
      def <init>() = {
        super.<init>;
        ()
      };
      val res1 = List("ab/bbc", "xxx/y", "ab/c").withFilter(((check$ifrefutable$1) => check$ifrefutable$1: @scala.unchecked match {
        case X((s @ _), (i @ _)) => true
        case _ => false
      })).map(((x$2) => x$2: @scala.unchecked match {
        case (x$1 @ X((s @ _), (i @ _))) => {
          implicit val t = s;
          scala.Tuple2(x$1, t)
        }
      })).map(((x$3) => {
        <synthetic> <artifact> private[this] val x$4 = x$3: @scala.unchecked: @scala.unchecked match {
          case scala.Tuple2(X((s @ _), (i @ _)), (t @ _)) => scala.Tuple3(s, i, t)
        };
        val s = x$4._1;
        val i = x$4._2;
        implicit val t = x$4._3;
        implicitly[String].+(i)
      }))
    }
  }
}
res1: List[String] = List(ab2, xxx3, ab2)
```

### CODE REVIEW

**Comment by retronym** [comment]
Any spec change required? 

There is probably a little work required in quasiquotes to bring them into line

**Comment by som-snytt** [comment]
There's the `implicit` part. I wanted to try something minimalist.

Actually, I want to fix SI-9460 by tupling the results of pattern matches instead of reapplying matches on values; currently, extractors are run multiple times, which isn't obvious or maybe widely known.

This PR is just a test; I'm working this weekend and won't drive it home straightaway.

I guess I'll try a clean build a bit later...

**Comment by som-snytt** [comment]
@retronym Is this worth following up with spec wording? I'll have to experiment if quasiquotes is broken. I definitely want to use the pun on Splenda and sugar.

File: src/reflect/scala/reflect/internal/TreeGen.scala [hunk 1]
```diff
@@ -656,36 +670,48 @@ abstract class TreeGen {
...
-          if (t.pos == NoPosition) NoPosition
-          else rangePos(t.pos.source, t.pos.start, t.pos.point, rhs1.pos.end)
-        val vfrom1 = ValFrom(atPos(wrappingPos(allpats)) { mkTuple(allpats) }, rhs1).setPos(pos1)
-        mkFor(vfrom1 :: rest1, sugarBody)
+      case (t @ ValFrom(mods, pat, rhs)) :: Nil =>
+        makeCombination(closurePos(t.pos), mapName, qual = rhs, mods, pat = pat, body = body)
+      case (t @ ValFrom(mods, pat, rhs)) :: (rest @ (ValFrom(_, _, _) :: _)) =>
+        makeCombination(closurePos(t.pos), flatMapName, qual = rhs, mods, pat = pat, body = mkFor(rest, sugarBody))
+      case (t @ ValFrom(mods, pat, rhs)) :: Filter(test) :: rest =>
+        val filtering = makeCombination(rhs.pos union test.pos, nme.withFilter, rhs, Modifiers(0), pat.duplicate, test)
```
Comment (som-snytt) [inline src/reflect/scala/reflect/internal/TreeGen.scala]: `val filtering = makeCombination(rhs.pos union test.pos, nme.withFilter, rhs, mods, pat.duplicate, test)`

**Comment by SethTisue** [comment]
at the moment the action on this is over at https://github.com/scala/slip/issues/6. @som-snytt, we won't look at the code in this PR in detail unless Martin+committee decide the SIP should become active (in which case yes, following up with spec wording, presumably adapted from @haroldl, will be key).

for now, our main question about the code is whether it corresponds to "the generalization" @retronym suggested at ~~https://github.com/scala/slip/issues/6~~ https://github.com/scala/scala.github.com/pull/417

**Comment by som-snytt** [comment]
@SethTisue No, it doesn't apply to patterns in general.

**Comment by retronym** [comment]
I'd suggest a test to involving a wildcard binder (`implicit _ <-` or `implicit (_, _) <-`.)  

Also something like:

``` scala
scala> for (/*implicit*/ Some(x) <- List(Some(1))) yield { val x = 2; x }
res8: List[Int] = List(2)
```

(I worry you're going to double define `x` in the block containing the yielded expression.)

**Comment by som-snytt** [comment]
That's a good thought. Saved by the brace.

```
      val res2 = List(Option(42)).withFilter(((check$ifrefutable$1) => check$ifrefutable$1: @scala.unchecked match {
        case Some((x @ _)) => true
        case _ => false
      })).map(((x$1) => {
        implicit val x = x$1: @scala.unchecked: @scala.unchecked match {
          case Some((x @ _)) => x
        };
        {
          val x = 17;
          x
        }
      }))
```

The fresh var is not implicit yet:

```
      val res8 = List(Option(42)).withFilter(((check$ifrefutable$1) => check$ifrefutable$1: @scala.unchecked match {
        case Some(_) => true
        case _ => false
      })).map(((x$1) => {
        <synthetic> <artifact> private[this] val x$2 = x$1: @scala.unchecked: @scala.unchecked match {
          case Some(_) => ()
        };
        {
          val x = 17;
          x
        }
      }))
```

**Comment by som-snytt** [comment]
@som-snytt do you plan on getting back to this? or do you want me to take over?

It would be nice to move on to other for comprehension improvements.

**Comment by adriaanm** [comment]
I suggest we close this PR and create a dedicated repo for S(L)IPly work in progress. Gotta keep 'em separated. @SethTisue, @som-snytt, what say you?

**Comment by som-snytt** [comment]
@adriaanm you mean you folks want to create a special repo and have other folks make PRs against it? Remind me, did you want me to flag it when you get a bit pedantic? But it's not useful having PRs just hang out either. OTOH, PRs need champions. I'm not totally sure people really care about -- what was this PR about again? -- oh yeah, introducing implicits in for expr bindings.

**Comment by adriaanm** [comment]
I'm thinking that repo could make it easier for people to track and cheer on language changes as opposed to the day-to-day already-agreed-on work. I want it to be clear which bars must be met by various PRs. 

I'm also fine with introducing a different milestone for PRs that are not yet ready for merging. It's my understanding this PR cannot be merged until the SIP committee agrees on the language change. That's a much higher bar that most other PRs targeting M4.

**Comment by som-snytt** [comment]
I should probably frequent the dive bar and avoid the high bar.

You're probably right that a SIP repo would be a place to gather. And it could have a gitter room.

**Comment by som-snytt** [comment]
Closing until someone takes me out for drinks.

**Comment by som-snytt** [comment]
Getting it off the queue.

**Pull request closed without merging**